### PR TITLE
refactor(tests): Refactor e2e tests entirely.

### PR DESCRIPTION
**Pull request by hansl**
More simple in the future, since things will be split in different files. The workflow follows the sort order of the files.

Also in this PR:
* Some fixes for typings when compiling the code.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -22,11 +24,18 @@ matrix:
   exclude:
     - node_js: "6"
       env: SCRIPT=lint
+    - node_js: "6"
```
Comment (filipesilva) [inline .travis.yml]: `SCRIPT=build` seems to be disabled, should this be here?

File: tests/e2e/000-setup/020-create-project.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      .then(() => process.chdir('./test-project'));
+  }
+
+  if (argv.nightly) {
+    // Install over the project with nightly builds.
+    const angularPackages = [
+      'core',
+      'common',
+      'compiler',
+      'http',
```
Comment (filipesilva) [inline tests/e2e/000-setup/020-create-project.ts]: `forms` is missing.

File: tests/e2e/100-build/200-base-href.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import {ng} from '../utils/process';
+import {expectFileToMatch} from '../utils/fs';
+
+
+export default function() {
+  return ng('build', '--base-href', '/myUrl')
```
Comment (filipesilva) [inline tests/e2e/100-build/200-base-href.ts]: Can we also test the default? I suppose it's not strictly needed but would be nice to check.

File: tests/e2e/100-build/200-environment.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import {ng} from '../utils/process';
+import {expectFileToMatch} from '../utils/fs';
+import {expectGitToBeClean} from '../utils/git';
+import {expectToFail} from '../utils/utils';
+
+
+export default function() {
+  // Try a prod build.
+  return ng('build', '--env=prod')
```
Comment (filipesilva) [inline tests/e2e/100-build/200-environment.ts]: Can we also test `--env=dev`?

File: tests/e2e/100-build/300-no-implicit-any.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+import {ng} from '../utils/process';
+import {gitClean} from '../utils/git';
+
+
+export default function() {
+  return updateTsConfig(json => {
+    json['compilerOptions']['noImplicitAny'] = true;
+  })
+  .then(() => ng('build'))
+  .then(() => gitClean());
```
Comment (filipesilva) [inline tests/e2e/100-build/300-no-implicit-any.ts]: What is the purpose of `gitClean` here? Other tests that build aren't cleaning afterwards.

File: tests/e2e/100-build/500-styles/500-css.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import {writeMultipleFiles, expectFileToMatch} from '../../utils/fs';
+import {ng} from '../../utils/process';
+
+
+export default function() {
+  return writeMultipleFiles({
+    'src/styles.css': `
```
Comment (filipesilva) [inline tests/e2e/100-build/500-styles/500-css.ts]: `500-styles/500-x` tests should probably be `501-x` etc.

File: tests/e2e/200-test/100-e2e.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  return ngServe(...args)
+    .then(() => ng('e2e'))
+    .then(() => killAllProcesses(), (err: any) => {
+      killAllProcesses();
+      throw err;
+    });
+}
+
+export default function() {
+  // This is supposed to fail with serving first...
```
Comment (filipesilva) [inline tests/e2e/200-test/100-e2e.ts]: `with`->`without`

File: tests/e2e/utils/project.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+
+export function updateTsConfig(fn: (json: any) => any | void) {
+  return updateJsonFile(tsConfigPath, fn);
+}
+
+
+export function ngServe(...args: string[]) {
+  return execAndWaitForOutputToMatch('ng', ['serve', ...args], /webpack: bundle is now VALID/);
```
Comment (filipesilva) [inline tests/e2e/utils/project.ts]: 👍

File: tests/e2e/100-build/500-styles/500-css.ts (hunk 6 above)
Comment (hansl) [inline tests/e2e/100-build/500-styles/500-css.ts]: I wanted to split them into their own directory.

File: tests/e2e/100-build/200-environment.ts (hunk 4 above)
Comment (hansl) [inline tests/e2e/100-build/200-environment.ts]: in `dev-build`

File: tests/e2e/100-build/200-base-href.ts (hunk 3 above)
Comment (hansl) [inline tests/e2e/100-build/200-base-href.ts]: What default?  No argument?

File: tests/e2e/000-setup/020-create-project.ts (hunk 2 above)
Comment (hansl) [inline tests/e2e/000-setup/020-create-project.ts]: done.

File: .travis.yml (hunk 1 above)
Comment (hansl) [inline .travis.yml]: Yes, when we re-enable it.

File: tests/e2e/200-test/100-e2e.ts (hunk 7 above)
Comment (hansl) [inline tests/e2e/200-test/100-e2e.ts]: Done.

File: tests/e2e/100-build/200-base-href.ts (hunk 3 above)
Comment (filipesilva) [inline tests/e2e/100-build/200-base-href.ts]: Yeah, that not adding a base href will use the default `/`.

**Comment by filipesilva** [comment]
`tests/e2e/utils/ng.ts` is an empty file, maybe you wanted to delete it?

File: tests/e2e_runner.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    console.error(red(err.message));
+    console.error(red(err.stack));
+
+    if (argv.debug) {
+      console.log(`Current Directory: ${process.cwd()}`);
+      console.log('Will loop forever while you debug... CTRL-C to quit.');
+
+      /* eslint-disable no-constant-condition */
+      while (1) {
+        // That's right!
```
Comment (filipesilva) [inline tests/e2e_runner.js]: /giphy oh-you

**Pull request merged**